## 1. 프록시와 내부 호출 문제
AOP를 적용하면 스프링은 대상 객체 대신 프록시를 스프링 빈으로 등록하기 때문에, 스프링은 의존관계 주입시에 항상 프록시 객체를 주입한다.
그러나 해당 프록시 객체의 메소드에서 처리가 종결되지않고, 
대상 객체의 내부에서 메서드 호출이 발생하면 프록시를 거치지 않고 대상 객체를 직접 호출하는 문제가 발생한다.
```java
@Slf4j
@Component
public class CallServiceV0 {

	public void external() {
		log.info("call external method");
		// 내부 메서드 호출
		this.internal();
	}
	
	public void internal() {
		log.info("call internal method");
	}
}
```
아래의 로그는 internal()과 external()모두 aop={}를 출력하는 AOP가 걸려있지만 external() 호출시 internal()에는 AOP가 걸려있지 않음을 확인할 수 있다.
internal()의 호출처는 실제 인스턴스 this(자기자신)이기 때문
```
2022-03-27 12:48:17.651  INFO 10948 --- [           main] c.e.d.o.a.p.i.CallServiceV0Test          : [test external]target=class com.example.demo.order.aop.problems.internalcall.CallServiceV0$$EnhancerBySpringCGLIB$$fd8e1ab2
2022-03-27 12:48:17.655  INFO 10948 --- [           main] c.e.d.o.a.p.internalcall.CallLogAspect   : aop=void com.example.demo.order.aop.problems.internalcall.CallServiceV0.external()
2022-03-27 12:48:17.664  INFO 10948 --- [           main] c.e.d.o.a.p.internalcall.CallServiceV0   : call external method
2022-03-27 12:48:17.664  INFO 10948 --- [           main] c.e.d.o.a.p.internalcall.CallServiceV0   : call internal method
2022-03-27 12:48:17.674  INFO 10948 --- [           main] c.e.d.o.a.p.i.CallServiceV0Test          : [test internal]target=class com.example.demo.order.aop.problems.internalcall.CallServiceV0$$EnhancerBySpringCGLIB$$fd8e1ab2
2022-03-27 12:48:17.674  INFO 10948 --- [           main] c.e.d.o.a.p.internalcall.CallLogAspect   : aop=void com.example.demo.order.aop.problems.internalcall.CallServiceV0.internal()
2022-03-27 12:48:17.674  INFO 10948 --- [           main] c.e.d.o.a.p.internalcall.CallServiceV0   : call internal method
```
위와같이 프록시 방식의 AOP는 메소드 내부호출에 대해서까지는 프록시를 적용할 수 없는 한계점이 존재한다. *(대안은 존재함)*
